src/node/desktop/src/native/dock/dock_mac.mm

```diff
@@ -0,0 +1,320 @@
+/*
+ * dock_mac.mm
+ *
+ * Copyright (C) 2022 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+// NOTE: Whenever adding new methods to be exported here, be sure to update
+// 'dock.node.d.ts', to ensure that the JavaScript type definitions are
+// visible for usage on the Electron side.
+
+#import <napi.h>
+#import <Cocoa/Cocoa.h>
+
+#define RS_EXPORT_FUNCTION(__NAME__, __FUNCTION__) \
+  exports.Set(                                     \
+    Napi::String::New(env, __NAME__),              \
+    Napi::Function::New(env, __FUNCTION__)         \
+  )
+
+// Implementation based on https://github.com/jessesquires/JSCustomBadge, and
+// https://cgit.kde.org/kdevplatform.git/commit/?id=c0626300e1c18448dd653a3484fd1ca6298e2d8c
+
+@interface MacDockTileView : NSView {
```

Comment:
Ah, I have a soft spot for Objective-C. Was doing NeXTSTEP programming in it long, long ago.